File: clients/typescript/src/models/turn-response.ts
/**
 * {@link TurnResponse} — the `application/json` body of a non-streaming turn
 * (`POST /v1/conversations/{id}/turns`, `POST /v1/turns`).
 *
 * `cost` is Loom's authoritative priced {@link TurnCost} for this turn —
 * computed once, inline, at turn time — or `null` when no price is configured
 * for the turn's `(provider, model)`; a pricing miss never fails the turn. The
 * streaming counterpart carries the identical value on the terminal
 * `turn_ended` event's `cost` field (see {@link TurnEventKind}), so
 * {@link collect} on a streamed turn agrees with this envelope for the same
 * input. Parsed with Zod as a response body.
 */

import { z } from "zod";

import type { DeepReadonly } from "../deep-readonly.types.js";
import { messageSchema } from "./message.js";
import { turnCostSchema } from "./turn-cost.js";

/** The `application/json` body of a non-streaming turn. */
export const turnResponseSchema = z.object({
  message: messageSchema,
  cost: turnCostSchema.nullable(),
});

/** The `application/json` body of a non-streaming turn. */
export type TurnResponse = DeepReadonly<z.infer<typeof turnResponseSchema>>;

File: crates/loom-server/src/v1/turn_response.rs
//! The non-streaming turn response envelope, [`TurnResponse`].

use serde::Serialize;
use utoipa::ToSchema;

use loom_core::Message;
use loom_provider::TurnCost;

/// The `application/json` body returned by a non-streaming turn
/// (`POST /v1/conversations/{id}/turns`, `POST /v1/turns`).
///
/// `cost` is Loom's authoritative priced cost for this turn — computed once,
/// inline, from the effective price for the turn's `(provider, model)` at turn
/// time, not `GET /v1/usage`'s eventually-consistent, asynchronously drained
/// aggregate. It is `None` only when no price is configured for the
/// (provider, model); a pricing miss never fails the turn. The streaming
/// counterpart (`stream=true`) carries the identical value on the terminal
/// `turn_ended` [`TurnEvent`](loom_provider::TurnEvent)'s `cost` field, so a
/// caller sees the same number whether it streams or not.
#[derive(Debug, Serialize, ToSchema)]
pub(super) struct TurnResponse {
    /// The assistant's turn.
    pub(super) message: Message,
    /// Loom's authoritative priced cost for this turn, or `None` when no price
    /// is configured for the (provider, model).
    pub(super) cost: Option<TurnCost>,
}
